common/src/resources/range_map.rs

```diff
@@ -0,0 +1,236 @@
+use std::{collections::HashMap, ops::Range};
+
+use crate::ether::evm::core::opcodes::WrappedOpcode;
+
+#[derive(Copy, Clone, Debug)]
+enum CollisionKind {
+    Deletion,
+    Splitting,
+    Shortening,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RangeMap(pub HashMap<Range<usize>, WrappedOpcode>);
+
+impl RangeMap {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Given an offset into memory, returns the associated opcode if it exists
+    pub fn get_by_offset(&self, offset: usize) -> Option<WrappedOpcode> {
+        self.0.get(self.find_range(offset).expect("RangeMap::have_range is broken")).cloned()
+    }
+
+    /// Associates the provided opcode with the range of memory modified by writing a `size`-byte
+    /// value to `offset`.
+    ///
+    /// This range is exactly `[offset, offset + size - 1]`. This function ensures that any existing
+    /// ranges that our new range would collide with are dealt with accordingly, that is:
+    ///
+    ///  - deleted, if our range completely overwrites it,
+    ///  - split, if our range overwrites a subset that partitions it,
+    ///  - shortened, if our range overwrites such that only one "end" of it is overwritten
+    pub fn write(&mut self, offset: usize, size: usize, opcode: WrappedOpcode) {
+        let range: Range<usize> = Range { start: offset, end: size - 1 };
```

Comment:
```suggestion
        let range: Range<usize> = Range { start: offset, end: offset + size - 1 };
```

Reply:
Again, good catch, will need to patch tests (this will block merging obviously).